python_modules/dagster/dagster_tests/storage_tests/test_event_log.py

```diff
@@ -326,3 +330,48 @@ def _ops():
         )
 
     assert incremental_run_stats == run_stats
+
+
+def test_step_stats():
+    @op
+    def op_success(_):
+        return 1
+
+    @op
+    def asset_op(_):
+        yield AssetMaterialization(asset_key=AssetKey("asset_1"))
+        yield Output(1)
+
+    @op(out=Out(str))
+    def op_failure(_):
+        time.sleep(0.001)
+        raise RetryRequested(max_retries=3)
+
+    def _ops():
+        op_success()
+        asset_op()
+        op_failure()
+
+    events, result = _synthesize_events(_ops, check_success=False)
+
+    step_stats = build_run_step_stats_from_events(result.run_id, events)
+    assert len(step_stats) == 3
+    assert len([step for step in step_stats if step.status == StepEventStatus.SUCCESS]) == 2
+    assert len([step for step in step_stats if step.status == StepEventStatus.FAILURE]) == 1
+    assert all([step.run_id == result.run_id for step in step_stats])
+
+    op_failure_stats = next(
+        iter([step for step in step_stats if step.step_key == "op_failure"]), None
+    )
+    assert op_failure_stats
+    assert op_failure_stats.attempts == 4
+    assert len(op_failure_stats.attempts_list) == 4
+
+    # build up run stats through incremental events
+    incremental_step_stats = None
+    for event in events:
+        incremental_step_stats = build_run_step_stats_from_events(
+            result.run_id, [event], incremental_step_stats
+        )
```

Comment:
should we do any checkups of the partially constructed stats partway through to make sure they match what we expect? like when its picked up a marker start but not yet a marker end? or is the end part where it verifies the final results are correct sufficient?

Reply:
I thought about it, but I'm not sure it really matters?

Reply:
Okay, scratch that... one major consumer of step stats is the k8s executor which uses the presence of step stats as a proxy for started steps.  I updated the PR to revert back to the original behavior and created a helper function to power the partial snapshots.

Also converted the NT => record.